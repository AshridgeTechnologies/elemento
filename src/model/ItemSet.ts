import Element from './Element'
import BaseElement, {propDef} from './BaseElement'
import {ComponentType, ElementType, eventAction, PropertyDef, PropertyExpr, PropertyValueType, StylingProps} from './Types'
import {elementHasParentTypeOf} from './createElement'

const selectableChoices = ['none', 'single', 'multiple', 'multipleAuto'] as const
type Selectable = typeof selectableChoices[number]

type Properties = Partial<Readonly<{
    items: PropertyValueType<any[]>,
    selectedItems: PropertyValueType<any>,
    selectable: PropertyValueType<Selectable>,
    selectAction: PropertyExpr,
    itemStyles: StylingProps
}>>

export default class ItemSet extends BaseElement<Properties> implements Element {

    readonly kind = 'ItemSet'
    get iconClass() { return 'list_alt' }
    type(): ComponentType { return 'statefulUI' }

    get items() { return this.properties.items }
    get selectedItems() { return this.properties.selectedItems }
    get selectable() { return this.properties.selectable }
    get selectAction() { return this.properties.selectAction }
    get itemStyles() { return this.properties.itemStyles }

    get propertyDefs(): PropertyDef[] {
        return [
            propDef('items', 'expr', {state: true}),
            propDef('selectedItems', 'expr', {state: true}),
            propDef('selectable', selectableChoices, {state: true}),
            propDef('selectAction', eventAction('$item', '$itemId'), {state: true}),
            propDef('itemStyles', 'styles')
        ]
    }

    get stateProperties(): string[] {
        return super.stateProperties.concat([
            'items', 'selectable', 'selectedItems', 'selectedItem'
        ])
    }

    canContain(elementType: ElementType) {
        return elementHasParentTypeOf(elementType, this)
    }
}
