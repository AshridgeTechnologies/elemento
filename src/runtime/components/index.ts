export {default as App} from './App'
export {default as Button} from './Button'
export {default as Icon} from './Icon'
export {default as Image} from './Image'
export {default as UserLogon} from './UserLogon'
export {default as Menu} from './Menu'
export {default as MenuItem} from './MenuItem'
export {default as NumberInput} from './NumberInput'
export {default as DateInput} from './DateInput'
export {default as TrueFalseInput} from './TrueFalseInput'
export {default as SpeechInput} from './SpeechInput'
export {default as Data} from './Data'
export {default as Calculation} from './Calculation'
export {default as Timer} from './Timer'
export {default as Collection} from './Collection'
export {default as Page} from './Page'
export {default as ComponentElement} from './ComponentElement'
export {default as Form} from './Form'
export {default as BaseFormState} from './FormState'
export {default as SelectInput} from './SelectInput'
export {default as TextElement} from './TextElement'
export {default as Layout} from './Layout'
export {default as Block} from './Block'
export {default as AppBar} from './AppBar'
export {default as ListElement} from './ListElement'
export {default as ListItem} from './ListItem'
export {default as ItemSet} from './ItemSet'
export {default as ItemSetItem} from './ItemSetItem'
export {default as TextInput} from './TextInput'
export {default as MemoryDataStore} from './MemoryDataStore'
export {default as FileDataStore} from './FileDataStore'
export {default as FileDataStoreImpl} from './FileDataStoreImpl'
export {default as IdbDataStoreImpl} from './IdbDataStoreImpl'
export {default as BrowserDataStore} from './BrowserDataStore'
export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as FirestoreDataStoreImpl} from './FirestoreDataStoreImpl'
export {default as ServerAppConnector} from './ServerAppConnector'
export {default as Adapter} from './Adapter'
