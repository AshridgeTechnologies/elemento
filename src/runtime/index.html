<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="initial-scale=1, width=device-width" />
  <title>Elemento App</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap"/>
</head>
<body>
  <div id="main"></div>
  <script src="https://unpkg.com/react@17/umd/react.development.js" crossorigin></script>
  <script src="https://unpkg.com/react-dom@17/umd/react-dom.development.js" crossorigin></script>
  <script src="runApp.ts" type="module"></script>
</body>
</html>