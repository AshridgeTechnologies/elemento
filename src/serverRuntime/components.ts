export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as Collection} from './Collection'

// NOTE: Must add the name of each component to ./names.ts