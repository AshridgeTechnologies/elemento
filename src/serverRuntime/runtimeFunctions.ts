export const asCurrentUser = (user: {email: string, name: string}) => {
   return user && {...user, Name: user.name, Email: user.email}
}

export function codeGenerationError(_expr: string, _err: string) {
   return undefined
}

