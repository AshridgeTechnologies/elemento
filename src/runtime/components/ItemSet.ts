import React, {Fragment, MouseEventHandler, SyntheticEvent} from 'react'
import {asArray, indexedPath, lastItemIdOfPath, PropVal, StylesPropVals, valueOf, valueOfOneLevel, valueOfProps} from '../runtimeFunctions'
import {useGetObjectState} from '../appData'
import {BaseComponentState, ComponentState} from './ComponentState'
import {isNil, last, range, reverse, without} from 'ramda'
import {unique} from '../../util/helpers'
import {isNumeric} from 'validator'

const selectableChoices = ['none', 'single', 'multiple', 'multipleAuto'] as const
type Selectable = typeof selectableChoices[number]

type SelectionType = 'replace' | 'addRemove' | 'fromLast'

type OnClickFn = MouseEventHandler<HTMLDivElement>

type Properties = Readonly<{
    path: string,
    itemContentComponent: (props: { path: string, $item: any, $itemId: (string | number), $selected: boolean, onClick: OnClickFn }) => React.ReactElement | null,
    itemStyles?: StylesPropVals
}>

type StateProperties = Partial<Readonly<{
    items: PropVal<any[]>,
    selectable: PropVal<Selectable>,
    selectedItems: PropVal<any>,
    selectAction: ($items: any, $itemIds: any) => void}
>>

type StateUpdatableProperties = Partial<Readonly<{
    selectedItemIds: (string|number)[]
}
>>

const ItemSet = React.memo(function ItemSet({path, itemContentComponent, ...props}: Properties) {
    const state = useGetObjectState<ItemSetState>(path)
    const onClick: OnClickFn = (event:SyntheticEvent) => {
        const {shiftKey, ctrlKey, metaKey} = (event.nativeEvent as MouseEvent)
        if (shiftKey) {
            document.getSelection()?.empty()
        }
        const targetId = (event.target as HTMLElement).id
        const itemId = lastItemIdOfPath(targetId)
        if (itemId) {
            const selectionType: SelectionType = (metaKey || ctrlKey) ? 'addRemove' : shiftKey ? 'fromLast' : 'replace'
            state.onSelect(itemId, selectionType)
        }
    }

    const children = state.items.map((item, index) => {
            const itemId = item?.id ?? String(index)
            const itemPath = indexedPath(path, itemId)
            const selected = state.isSelected(index)
            return React.createElement(itemContentComponent, {path: itemPath, $item: item, $itemId: itemId, $selected: selected, onClick, key: itemId})
        }
    )

    return React.createElement(Fragment, null, ...children)
})

export default ItemSet

export class ItemSetState extends BaseComponentState<StateProperties, StateUpdatableProperties>
    implements ComponentState<ItemSetState>{

    get items(): any[] {
        return valueOfOneLevel<any>(this.props.items) ?? []
    }

    get selectable() {
        return valueOf(this.props.selectable) ?? 'single'
    }

    get selectAction() {
        return this.props.selectAction
    }

    get selectedItems(): any[] {
        const selectedItems = this.state.selectedItemIds !== undefined ? this.state.selectedItemIds : this.props.selectedItems
        const uniqueSelectedItems = unique(asArray(valueOfOneLevel(selectedItems)))
        return uniqueSelectedItems.map( id => this.findItem(id) ).filter( it => it !== undefined)
    }

    get selectedItemIds(): (string | number)[] {
        return this.state.selectedItemIds ?? unique(asArray(valueOfOneLevel(this.props.selectedItems))).map( it => this.findId(it)).filter( it => it !== undefined)
    }

    get selectedItem() {
        return last(this.selectedItems) ?? null
    }

    _setSelectedItems(selectedItems: any) {
        const selectedItemIds = selectedItems ? asArray(valueOfOneLevel(selectedItems)).map( (it: any) => this.findId(it)) : undefined
        this.latest().updateState({selectedItemIds})
    }

    Reset() {
        this._setSelectedItems(undefined)
    }

    Set(selectedItems: any) {
        this._setSelectedItems(selectedItems)
    }

    Select(selectedItems: any) {
        this._setSelectedItems(this.selectedItems.concat(valueOfOneLevel(selectedItems)))
    }

    isSelected(itemIndex: number) {
        const {selectedItemIds} = this
        const item = this.items[itemIndex]
        return selectedItemIds.includes(item?.id) || selectedItemIds.includes(itemIndex)
    }

    onSelect(selectedItemId: (string|number), type: SelectionType) {
        const selectActionSingle = () => {
            const newItemId = this.findId(selectedItemId)
            const newItem = this.findItem(newItemId)
            this.selectAction?.(newItem, newItemId)
        }

        if (this.selectable === 'none') {
            selectActionSingle()
        }
        if (this.selectable === 'single') {
            this._setSelectedItems(selectedItemId)
            selectActionSingle()
        }
        if (this.selectable === 'multiple' || this.selectable === 'multipleAuto') {
            if (type === 'addRemove' || this.selectable === 'multipleAuto') {
                const newItemId = this.findId(selectedItemId)
                const existingSelectedIds = this.selectedItemIds
                if (existingSelectedIds.includes(newItemId)) {
                    const newItemIds = without([newItemId], existingSelectedIds)
                    this._setSelectedItems(newItemIds)
                    this.selectAction?.(null, null)
                } else {
                    this._setSelectedItems(existingSelectedIds.concat(newItemId))
                    const newItemIds = [newItemId]
                    const newItems = newItemIds.map( it => this.findItem(it))
                    this.selectAction?.(newItems, newItemIds)
                }
            } else if (type === 'fromLast') {
                const itemIndex = (id: typeof selectedItemId) => {
                    const index = this.items.findIndex( it => it?.id === id )
                    if (index !== -1) {
                        return index
                    }
                    const numberIndex = Number(id)
                    if (!isNaN(numberIndex)) {
                        return numberIndex
                    }
                    return null
                }

                const fromIndex = itemIndex(last(this.selectedItemIds) ?? 0)
                const toIndex = itemIndex(selectedItemId)
                if (fromIndex !== null && toIndex !== null) {
                    const selectedIndexes = (toIndex >= fromIndex) ? range(fromIndex, toIndex + 1) : reverse(range(toIndex, fromIndex + 1))
                    this._setSelectedItems(selectedIndexes)
                    const newItemIds = selectedIndexes.map( ind => this.findId(ind) )
                    const newItems = newItemIds.map( it => this.findItem(it))
                    this.selectAction?.(newItems, newItemIds)
                }
            } else {
                this._setSelectedItems(selectedItemId)
                selectActionSingle()
            }
        }
    }

    private findItem(item: any) {
        if (typeof item === 'number' || typeof(item) === 'string' && isNumeric(item)) {
            return this.items[Number(item)]
        }
        return this.items.find( it => !isNil(it) && (it === item || it.id === item) ) ?? undefined
    }

    private findId(item: any) {
        const itemIndex = this.items.findIndex( it => !isNil(it) && (it === item || it.id === item) )
        if (itemIndex !== -1) {
            return this.items[itemIndex].id ?? itemIndex
        }

        const lookupItem = this.items[item]
        if (lookupItem !== undefined) {
            return lookupItem.id ?? Number(item)
        }

        return undefined
    }
}

(ItemSet as any).State = ItemSetState
