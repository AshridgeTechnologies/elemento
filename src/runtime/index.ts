import * as components from './components'
import {globalFunctions} from './globalFunctions'
export {components, globalFunctions}

export {default as React} from 'react'
export {default as ReactDOM} from 'react-dom/client'
export {default as appFunctions} from './appFunctions'
export {useGetObjectState, useObjectState} from './appData'
export {asArray, codeGenerationError, parentPath} from './runtimeFunctions'
export {default as AppRunner} from './AppRunner'
export {run} from './run'
