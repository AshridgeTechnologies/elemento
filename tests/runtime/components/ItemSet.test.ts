/**
 * @jest-environment jsdom
 */

import {createElement, Fragment, MouseEventHandler} from 'react'
import {ItemSet, TextElement} from '../../../src/runtime/components/index'
import {snapshot, wait, wrappedTestElement} from '../../testutil/testHelpers'
import {actWait, testContainer} from '../../testutil/rtlHelpers'
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom'
import {ItemSetState} from '../../../src/runtime/components/ItemSet'
import {highlightClassName, highlightElement, StylesProps} from '../../../src/runtime/runtimeFunctions'
import renderer from 'react-test-renderer'
import {ItemSetItem} from '../../../src/runtime/components'

function ItemSetItem1(props: {path: string, $item: {text: string}, $selected: boolean, onClick: MouseEventHandler<HTMLDivElement>, styles: StylesProps}) {
    const styles = {color: 'red', width: 300}
    return createElement(ItemSetItem, {path: props.path, onClick: props.onClick, styles, children:
        createElement(TextElement, {path: `${props.path}.Text99`}, props.$item.text, 'selected', props.$selected.toString()) }
    )
}

const itemSetData = [{id: 'id1', text: 'where are you?'}, {id: 'id2', text: 'over here!'}]
const longItemSetData = [{id: 'id1', text: 'One'}, {id: 'id2', text: 'Two'}, {id: 'id3', text: 'Three'}, {id: 'id4', text: 'Four'}]
const itemSetDataNoIds = [{Label: 'No 1', text: 'where are you?'}, {Label: 'No 2', text: 'over here!'}]

const [itemSet, appStoreHook] = wrappedTestElement(ItemSet, ItemSetState)

const stateAt = (path: string) => appStoreHook.stateAt(path)

test('ItemSet produces output containing ReactElement children', () => {
    snapshot(itemSet('app.page1.itemSet1', {items: itemSetData}, {itemContentComponent: ItemSetItem1, itemStyles: {color: 'red', width: 300}}))()
})

test('ItemSet produces output with indexes if items have no ids', () => {
    snapshot(itemSet('app.page1.itemSet1', {items: itemSetDataNoIds}, {itemContentComponent: ItemSetItem1, itemStyles: {color: 'red', width: 200}}))()
})

test('ItemSet shows single selectedItem as selected', async () => {
    snapshot(itemSet('app.page1.itemSet1', {selectedItems: itemSetData[1], items: itemSetData}, {itemContentComponent: ItemSetItem1}))()
})

test('ItemSet shows single selectedItem as selected if no ids', () => {
    snapshot(itemSet('app.page1.itemSet1', {selectedItems: itemSetDataNoIds[1], items: itemSetDataNoIds}, {itemContentComponent: ItemSetItem1}))()
})

test('ItemSet shows only first selectedItem as selected if no ids when given zero as index', async () => {
    let el: any
    await actWait(async () => {
        el = renderer.create(itemSet('app.page1.itemSet1', {selectedItems: 0, items: itemSetDataNoIds}, {
            itemContentComponent: ItemSetItem1
        }))
    })

    expect(el.toJSON()).toMatchSnapshot()
})

test('ItemSet produces empty output if items is null but there is a selected item, as may happen when refreshing data', () => {
    snapshot(itemSet('app.page1.itemSet1', {selectedItem: itemSetData[1], items: null}, {itemContentComponent: ItemSetItem1}))()
})

test('ItemSet updates its selectedItem in the app state', async () => {
    const {el, user}  = testContainer(itemSet('app.page1.itemSet1', {items: itemSetData}, {itemContentComponent: ItemSetItem1}))
    const itemSetItem0El = el`[id="app.page1.itemSet1.#id1.Text99"]`
    await user.click(itemSetItem0El)
    expect(stateAt('app.page1.itemSet1').selectedItem).toStrictEqual(itemSetData[0])
})

test('ItemSet adds to selectedItems in the app state with Control key', async () => {
    const {el, user}  = testContainer(itemSet('app.page1.itemSet1', {items: longItemSetData, selectedItems: 1, selectable: 'multiple'}, {itemContentComponent: ItemSetItem1}), 'container2')
    const itemSetItem0El = el`[id="app.page1.itemSet1.#id1.Text99"]`
    await user.keyboard('[ControlLeft>]')
    await user.click(itemSetItem0El)
    expect(stateAt('app.page1.itemSet1').selectedItems).toStrictEqual([longItemSetData[1], longItemSetData[0]])
})

test('ItemSet selects a block in the app state with Shift key', async () => {
    const {el, user}  = testContainer(itemSet('app.page1.itemSet1', {items: longItemSetData, selectedItems: 2, selectable: 'multiple'}, {itemContentComponent: ItemSetItem1}), 'container3')
    const itemSetItem0El = el`[id="app.page1.itemSet1.#id1.Text99"]`
    await user.keyboard('[ShiftLeft>]')
    await user.click(itemSetItem0El)
    expect(stateAt('app.page1.itemSet1').selectedItems).toStrictEqual([longItemSetData[2], longItemSetData[1], longItemSetData[0]])
})

test('ItemSet updates the selectedItem when it changes', async () => {
    const {el, user, renderThe}  = testContainer(itemSet('app.page1.itemSet1', {items: itemSetData}, {itemContentComponent: ItemSetItem1}))
    const itemSetItem0El = el`[id="app.page1.itemSet1.#id1.Text99"]`
    await user.click(itemSetItem0El)
    expect(stateAt('app.page1.itemSet1').selectedItem).toStrictEqual(itemSetData[0])

    const updatedItemSetData = [{id: 'id1', text: 'new text'}, itemSetData[1]]
    renderThe(itemSet('app.page1.itemSet1', {items: updatedItemSetData}, {itemContentComponent: ItemSetItem1}))
    await wait(20)
    expect(stateAt('app.page1.itemSet1').selectedItem).toBe(updatedItemSetData[0])
})

test('ItemSet does not update its selectedItem if not selectable', async () => {
    let container = testContainer(itemSet('app.page1.itemSet1', {items: itemSetData, selectable: 'none'}, {itemContentComponent: ItemSetItem1}), 'container4')
    const itemSetItem0El = container.querySelector('[id="app.page1.itemSet1.#id1.Text99"]')
    const user = userEvent.setup()
    await user.click(itemSetItem0El)
    expect(stateAt('app.page1.itemSet1').selectedItem).toBe(null)
})

test('Can highlight all matching elements in a itemSet', async () => {
    const {el} = testContainer(itemSet('app.page1.itemSet1', {items: itemSetData}, {itemContentComponent: ItemSetItem1}))
    highlightElement('app.page1.itemSet1.Text99')

    const itemSetItem0El = el`[id="app.page1.itemSet1.#id1.Text99"]`
    const itemSetItem1El = el`[id="app.page1.itemSet1.#id2.Text99"]`
    expect(itemSetItem0El).toHaveClass(highlightClassName)
    expect(itemSetItem1El).toHaveClass(highlightClassName)
})
