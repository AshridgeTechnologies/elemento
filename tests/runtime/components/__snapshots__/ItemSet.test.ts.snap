// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`produces empty output if items is null but there is a selected item, as may happen when refreshing data 1`] = `
[
  <div
    id="DndDescribedBy-6"
    style={
      {
        "display": "none",
      }
    }
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>,
  <div
    aria-atomic={true}
    aria-live="assertive"
    id="DndLiveRegion-6"
    role="status"
    style={
      {
        "border": 0,
        "clip": "rect(0 0 0 0)",
        "clipPath": "inset(100%)",
        "height": 1,
        "margin": -1,
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "whiteSpace": "nowrap",
        "width": 1,
      }
    }
  />,
]
`;

exports[`produces output containing ReactElement children 1`] = `
[
  <div
    aria-describedby="DndDescribedBy-0"
    aria-disabled={true}
    aria-roledescription="draggable"
    className="MuiBox-root css-1r6lq6j"
    id="app.page1.itemSet1.#id1"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      id="app.page1.itemSet1.#id1.Text99"
    >
      where are you?
      <br />
      item id id1
      <br />
      index 0
      <br />
      selected
      <br />
      false
    </div>
  </div>,
  <div
    aria-describedby="DndDescribedBy-0"
    aria-disabled={true}
    aria-roledescription="draggable"
    className="MuiBox-root css-1r6lq6j"
    id="app.page1.itemSet1.#id2"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      id="app.page1.itemSet1.#id2.Text99"
    >
      over here!
      <br />
      item id id2
      <br />
      index 1
      <br />
      selected
      <br />
      false
    </div>
  </div>,
  <div
    id="DndDescribedBy-0"
    style={
      {
        "display": "none",
      }
    }
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>,
  <div
    aria-atomic={true}
    aria-live="assertive"
    id="DndLiveRegion-0"
    role="status"
    style={
      {
        "border": 0,
        "clip": "rect(0 0 0 0)",
        "clipPath": "inset(100%)",
        "height": 1,
        "margin": -1,
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "whiteSpace": "nowrap",
        "width": 1,
      }
    }
  />,
]
`;

exports[`produces output containing ReactElement children if draggable 1`] = `
[
  <div
    aria-describedby="DndDescribedBy-1"
    aria-disabled={false}
    aria-roledescription="draggable"
    className="MuiBox-root css-1r6lq6j"
    id="app.page1.itemSet1.#id1"
    onClick={[Function]}
    onKeyDown={[Function]}
    onPointerDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      id="app.page1.itemSet1.#id1.Text99"
    >
      where are you?
      <br />
      item id id1
      <br />
      index 0
      <br />
      selected
      <br />
      false
    </div>
  </div>,
  <div
    aria-describedby="DndDescribedBy-1"
    aria-disabled={false}
    aria-roledescription="draggable"
    className="MuiBox-root css-1r6lq6j"
    id="app.page1.itemSet1.#id2"
    onClick={[Function]}
    onKeyDown={[Function]}
    onPointerDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      id="app.page1.itemSet1.#id2.Text99"
    >
      over here!
      <br />
      item id id2
      <br />
      index 1
      <br />
      selected
      <br />
      false
    </div>
  </div>,
  <div
    id="DndDescribedBy-1"
    style={
      {
        "display": "none",
      }
    }
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>,
  <div
    aria-atomic={true}
    aria-live="assertive"
    id="DndLiveRegion-1"
    role="status"
    style={
      {
        "border": 0,
        "clip": "rect(0 0 0 0)",
        "clipPath": "inset(100%)",
        "height": 1,
        "margin": -1,
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "whiteSpace": "nowrap",
        "width": 1,
      }
    }
  />,
]
`;

exports[`produces output when an item is null 1`] = `
[
  <div
    aria-describedby="DndDescribedBy-7"
    aria-disabled={true}
    aria-roledescription="draggable"
    className="MuiBox-root css-1r6lq6j"
    id="app.page1.itemSet1.#id1"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      id="app.page1.itemSet1.#id1.Text99"
    >
      One
      <br />
      item id id1
      <br />
      index 0
      <br />
      selected
      <br />
      false
    </div>
  </div>,
  <div
    aria-describedby="DndDescribedBy-7"
    aria-disabled={true}
    aria-roledescription="draggable"
    className="MuiBox-root css-1r6lq6j"
    id="app.page1.itemSet1.#1"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      id="app.page1.itemSet1.#1.Text99"
    >
      <br />
      item id 1
      <br />
      index 1
      <br />
      selected
      <br />
      false
    </div>
  </div>,
  <div
    aria-describedby="DndDescribedBy-7"
    aria-disabled={true}
    aria-roledescription="draggable"
    className="MuiBox-root css-1r6lq6j"
    id="app.page1.itemSet1.#id3"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      id="app.page1.itemSet1.#id3.Text99"
    >
      Three
      <br />
      item id id3
      <br />
      index 2
      <br />
      selected
      <br />
      false
    </div>
  </div>,
  <div
    id="DndDescribedBy-7"
    style={
      {
        "display": "none",
      }
    }
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>,
  <div
    aria-atomic={true}
    aria-live="assertive"
    id="DndLiveRegion-7"
    role="status"
    style={
      {
        "border": 0,
        "clip": "rect(0 0 0 0)",
        "clipPath": "inset(100%)",
        "height": 1,
        "margin": -1,
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "whiteSpace": "nowrap",
        "width": 1,
      }
    }
  />,
]
`;

exports[`produces output with indexes if items have no ids 1`] = `
[
  <div
    aria-describedby="DndDescribedBy-2"
    aria-disabled={true}
    aria-roledescription="draggable"
    className="MuiBox-root css-1r6lq6j"
    id="app.page1.itemSet1.#0"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      id="app.page1.itemSet1.#0.Text99"
    >
      where are you?
      <br />
      item id 0
      <br />
      index 0
      <br />
      selected
      <br />
      false
    </div>
  </div>,
  <div
    aria-describedby="DndDescribedBy-2"
    aria-disabled={true}
    aria-roledescription="draggable"
    className="MuiBox-root css-1r6lq6j"
    id="app.page1.itemSet1.#1"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      id="app.page1.itemSet1.#1.Text99"
    >
      over here!
      <br />
      item id 1
      <br />
      index 1
      <br />
      selected
      <br />
      false
    </div>
  </div>,
  <div
    id="DndDescribedBy-2"
    style={
      {
        "display": "none",
      }
    }
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>,
  <div
    aria-atomic={true}
    aria-live="assertive"
    id="DndLiveRegion-2"
    role="status"
    style={
      {
        "border": 0,
        "clip": "rect(0 0 0 0)",
        "clipPath": "inset(100%)",
        "height": 1,
        "margin": -1,
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "whiteSpace": "nowrap",
        "width": 1,
      }
    }
  />,
]
`;

exports[`shows items with no ids 1`] = `
[
  <div
    aria-describedby="DndDescribedBy-4"
    aria-disabled={true}
    aria-roledescription="draggable"
    className="MuiBox-root css-1r6lq6j"
    id="app.page1.itemSet1.#0"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      id="app.page1.itemSet1.#0.Text99"
    >
      where are you?
      <br />
      item id 0
      <br />
      index 0
      <br />
      selected
      <br />
      false
    </div>
  </div>,
  <div
    aria-describedby="DndDescribedBy-4"
    aria-disabled={true}
    aria-roledescription="draggable"
    className="MuiBox-root css-1r6lq6j"
    id="app.page1.itemSet1.#1"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      id="app.page1.itemSet1.#1.Text99"
    >
      over here!
      <br />
      item id 1
      <br />
      index 1
      <br />
      selected
      <br />
      true
    </div>
  </div>,
  <div
    id="DndDescribedBy-4"
    style={
      {
        "display": "none",
      }
    }
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>,
  <div
    aria-atomic={true}
    aria-live="assertive"
    id="DndLiveRegion-4"
    role="status"
    style={
      {
        "border": 0,
        "clip": "rect(0 0 0 0)",
        "clipPath": "inset(100%)",
        "height": 1,
        "margin": -1,
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "whiteSpace": "nowrap",
        "width": 1,
      }
    }
  />,
]
`;

exports[`shows only first selectedItem as selected if no ids when given zero as index 1`] = `
[
  <div
    aria-describedby="DndDescribedBy-5"
    aria-disabled={true}
    aria-roledescription="draggable"
    className="MuiBox-root css-1r6lq6j"
    id="app.page1.itemSet1.#0"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      id="app.page1.itemSet1.#0.Text99"
    >
      where are you?
      <br />
      item id 0
      <br />
      index 0
      <br />
      selected
      <br />
      true
    </div>
  </div>,
  <div
    aria-describedby="DndDescribedBy-5"
    aria-disabled={true}
    aria-roledescription="draggable"
    className="MuiBox-root css-1r6lq6j"
    id="app.page1.itemSet1.#1"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      id="app.page1.itemSet1.#1.Text99"
    >
      over here!
      <br />
      item id 1
      <br />
      index 1
      <br />
      selected
      <br />
      false
    </div>
  </div>,
  <div
    id="DndDescribedBy-5"
    style={
      {
        "display": "none",
      }
    }
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>,
  <div
    aria-atomic={true}
    aria-live="assertive"
    id="DndLiveRegion-5"
    role="status"
    style={
      {
        "border": 0,
        "clip": "rect(0 0 0 0)",
        "clipPath": "inset(100%)",
        "height": 1,
        "margin": -1,
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "whiteSpace": "nowrap",
        "width": 1,
      }
    }
  />,
]
`;

exports[`shows single selectedItem as selected 1`] = `
[
  <div
    aria-describedby="DndDescribedBy-3"
    aria-disabled={true}
    aria-roledescription="draggable"
    className="MuiBox-root css-1r6lq6j"
    id="app.page1.itemSet1.#id1"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      id="app.page1.itemSet1.#id1.Text99"
    >
      where are you?
      <br />
      item id id1
      <br />
      index 0
      <br />
      selected
      <br />
      false
    </div>
  </div>,
  <div
    aria-describedby="DndDescribedBy-3"
    aria-disabled={true}
    aria-roledescription="draggable"
    className="MuiBox-root css-1r6lq6j"
    id="app.page1.itemSet1.#id2"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      id="app.page1.itemSet1.#id2.Text99"
    >
      over here!
      <br />
      item id id2
      <br />
      index 1
      <br />
      selected
      <br />
      true
    </div>
  </div>,
  <div
    id="DndDescribedBy-3"
    style={
      {
        "display": "none",
      }
    }
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>,
  <div
    aria-atomic={true}
    aria-live="assertive"
    id="DndLiveRegion-3"
    role="status"
    style={
      {
        "border": 0,
        "clip": "rect(0 0 0 0)",
        "clipPath": "inset(100%)",
        "height": 1,
        "margin": -1,
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "whiteSpace": "nowrap",
        "width": 1,
      }
    }
  />,
]
`;
